task_processing/plugins/mesos/resource_helpers.py

```diff
@@ -0,0 +1,69 @@
+from typing import Any
+from typing import Dict
+
+from task_processing.plugins.mesos.task_config import MesosTaskConfig
+
+NUMERIC_RESOURCES = frozenset(['cpus', 'mem', 'disk', 'gpus'])
+OTHER_RESOURCES: dict = {'ports': []}
+
+
+def _make_resource_dict():
+    d = dict([(r, 0) for r in NUMERIC_RESOURCES])
+    d.update(OTHER_RESOURCES)
+    return d
+
+
+def get_offer_resources(offer, role: str) -> Dict[str, Any]:
+    """ Get the resources from a Mesos offer
+
+    :param offer: the payload from a Mesos resourceOffer call
+    :param role: the Mesos role we want to get resources for
+    :returns: a mapping from resource name -> available resources for the offer
+    """
+    offer_resources = _make_resource_dict()
+    for resource in offer.resources:
+        if resource.role != role:
+            continue
+
+        if resource.name in NUMERIC_RESOURCES:
+            offer_resources[resource.name] += resource.scalar.value
+        elif resource.name == 'ports':
+            offer_resources['ports'] = [r for r in resource.ranges.range]
+
+    return offer_resources
+
+
+def allocate_task_resources(task: MesosTaskConfig, offer_resources: Dict[str, Any]) -> int:
```

Comment:
I'd like to encourage you to not make this function rely on side effects, but just return a new dictionary that represents the offer_resource once the task config has been taken into account, and the caller can use that as they need to.

If we're going all in on the typing, can you also make the data structure that is returned by get_offer_resources something *real* (a named tuple, or a [PRecord](https://github.com/tobgu/pyrsistent#pr)) and update the type definition for the function. That way we can have more guarantees about the caller passing the right thing than those we get just asking them to read the docstring.

Finally it feels awkward that this returns a port (seemingly completely unrelated to whats hinted at by the function name). Can we separate the function that provides a new offer_resources with the numeric resources subtracted from that which knows how to deal with ports?

Reply:
btw, I kind of hinted at it, but this is exactly the kind of pattern that pyrsistent forces out of your code for you.You couldn't write this function using a immutable data structure for offer_resources.

Reply:
Fixed, see maksym's comment above.